### PR TITLE
Add `read` method to `StorageStreamDownloader`

### PR DESCRIPTION
**Pull request by jalauzon-msft**
This change adds a standard `read(size=-1)` method to the `StorageStreamDownloader` class to make it more like a Python stream. This method acts as one would expect, accepting a size to read and keeping track of a read offset internally. For now, the stream is non-seekable and forward reading only but I don't think there is anything preventing us from making this seekable in the future if desired. It will use an existing `_ChunkDownloader` to download chunks as needed based on the read size.

This change also adds support for async streams (classes with an async `read` method) to async `upload_blob`. This was needed to support using `StorageStreamDownloader` as an input to `upload_blob`.

### CODE REVIEW

**Review by vincenttran-msft — APPROVED** [review]
LGTM! Great work, this method looks quite hairy with that `parallel` stuff 😨

File: sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/uploads_async.py [hunk 1]
```diff
@@ -19,6 +19,26 @@
 from .uploads import SubStream, IterStreamer  # pylint: disable=unused-import
 
 
+async def _async_parallel_uploads(uploader, pending, running):
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_shared/uploads_async.py:22]: Note: This is a copy of the below method just modified to use `__anext__`

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared/uploads_async.py [hunk 2]
```diff
@@ -19,6 +19,26 @@
...
+            for _ in range(0, len(done)):
+                next_chunk = await pending.__anext__()
+                running.add(asyncio.ensure_future(uploader(next_chunk)))
+        except StopAsyncIteration:
+            break
+
+    # Wait for the remaining uploads to finish
+    if running:
+        done, _running = await asyncio.wait(running)
+        range_ids.extend([chunk.result() for chunk in done])
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_shared/uploads_async.py:38]: I'm assuming any errors raised from this get handled further up the stack?

File: sdk/storage/azure-storage-blob/azure/storage/blob/_download.py [hunk 3]
```diff
@@ -335,6 +335,7 @@ def __init__(
         self._non_empty_ranges = None
         self._response = None
         self._encryption_data = None
+        self._offset = 0
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_download.py:338]: Does this `offset` property need to be checked/respected by any of the other download APIs?
I'm curious what would happen for example, if someone called `download.read(10)` followed by `download.readall()`.
While I consider this unlikely - if would be good to check what happens to a standard file IO, or bytes IO if you try to do this and compare our behaviour.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_download.py [hunk 4]
```diff
@@ -552,6 +553,69 @@ def chunks(self):
...
+                stream=stream,
+                parallel=parallel,
+                validate_content=self._validate_content,
+                encryption_options=self._encryption_options,
+                encryption_data=self._encryption_data,
+                use_location=self._location_mode,
+                **self._request_options
+            )
+
+            if parallel:
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_download.py:601]: Maybe we should also predicate this on `remaining_size`? Not sure we'd want to start up a threadpool if `remaining_size` <= 1 chunk

**Review by annatisch — COMMENTED** [review]
I think this is looking really good - thanks so much for this!!

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared/uploads_async.py (hunk 2 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_shared/uploads_async.py:38]: I think so, yeah. This is pretty much a copy from the existing helper, just made to support an async generator which has existed for some time, so I have to assume error handling is correct :). Either way, this is all wrapped in try/except in the client that should handle any errors appropriately.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_download.py (hunk 4 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_download.py]: Good idea, done.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_download.py (hunk 3 above)
Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_download.py:337]: As discussed offline, we did need to handle `offset` in other read APIs to align with standard IO. I've done that in the latest iteration.

File: sdk/storage/azure-storage-blob/azure/storage/blob/aio/_download_async.py [hunk 5]
```diff
@@ -595,30 +610,36 @@ async def readinto(self, stream):
...
+        if remaining_size <= 0:
+            return 0
+
+        # Write the content to the user stream if there is data left
+        if self._offset < len(self._current_content):
+            content = self._current_content[self._offset:]
+            stream.write(content)
+            self._offset += len(content)
+            if self._progress_hook:
+                self._progress_hook(len(content), self.size)
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob/azure/storage/blob/aio/_download_async.py:625]: Shouldn't this be `await self._progress_hook(...)`?

Comment (jalauzon-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/aio/_download_async.py]: Good catch! Fixed.

**Pull request merged**